### PR TITLE
Workflow: Add bnb tests 

### PR DESCRIPTION
**Pull request by younesbelkada**
This PR adds a new workflow file in order to tests bnb pull requests easily

### CODE REVIEW

File: .github/workflows/tests-pr.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
+name: BNB from source self-hosted runner with slow tests (scheduled)
```
Comment (younesbelkada) [inline .github/workflows/tests-pr.yml]:
```suggestion
name: BNB tests on Pull Requests
```

File: .github/workflows/tests-pr.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  SLACK_API_TOKEN: ${{ secrets.SLACK_API_TOKEN }}
+  OUTPUT_SLACK_CHANNEL_ID: "C071049S8EM"
+
+jobs:
+  run_tests_single_gpu:
+    strategy:
+      fail-fast: false
+    runs-on: [self-hosted, single-gpu, nvidia-gpu, t4, ci]
+    container:
+      image: "huggingface/peft-gpu-bnb-latest:latest"
```
Comment (younesbelkada) [inline .github/workflows/tests-pr.yml:22]: we use this image that contains all HF libs installed from latest pypi, we already test bnb on HF libs from source on daily basis

File: .github/workflows/tests-pr.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      options: --gpus all --shm-size "16gb" --ipc host -v /mnt/cache/.cache/huggingface:/mnt/cache/
+    defaults:
+      run:
+        shell: bash
+    steps:
+      - uses: actions/checkout@v3
+      - name: Pip install and checkout
+        run: |
+          source activate peft
```
Comment (younesbelkada) [inline .github/workflows/tests-pr.yml:31]: `peft` is the name of the venv in the docker image

File: .github/workflows/tests-pr.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+          pip uninstall bitsandbytes
+          git remote add upstream https://github.com/TimDettmers/bitsandbytes.git
+          git fetch upstream pull/${{ github.event.inputs.pr_number }}/head:test-pr
+          git checkout test-pr
+          cmake -B . -DCOMPUTE_BACKEND=cuda -S . && cmake --build . && pip install -e . && pip freeze | grep bitsandbytes
```
Comment (younesbelkada) [inline .github/workflows/tests-pr.yml:36]: this is the command to build bnb from source

File: .github/workflows/tests-pr.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+          pip install pytest-reportlog pytest-cov parameterized datasets scipy einops
+          pip install "pytest>=7.2.0,<8.0.0" # see: https://github.com/huggingface/transformers/blob/ce4fff0be7f6464d713f7ac3e0bbaafbc6959ae5/setup.py#L148C6-L148C26
+      
+      - name: Run core tests on single GPU
+        id: tests
+        if: always()
+        run: |
+          source activate peft
+          pytest tests/
```
Comment (younesbelkada) [inline .github/workflows/tests-pr.yml:47]: this is to be adapted to add custom tests

File: .github/workflows/tests-pr.yml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+          
+      - name: Post to Slack
+        if: always()
+        uses: https://github.com/huggingface/transformers/blob/main/.github/actions/post-slack/action.yml@main
```
Comment (younesbelkada) [inline .github/workflows/tests-pr.yml:51]: this action is a simple GHA stored on transformers to post messages on slack

**Comment by younesbelkada** [comment]
cc @muellerzr

File: .github/workflows/tests-pr.yml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        run: |
+          source activate peft
+          pip uninstall bitsandbytes
+          git remote add upstream https://github.com/TimDettmers/bitsandbytes.git
+          git fetch upstream pull/${{ github.event.inputs.pr_number }}/head:test-pr
+          git checkout test-pr
+          cmake -B . -DCOMPUTE_BACKEND=cuda -S . && cmake --build . && pip install -e . && pip freeze | grep bitsandbytes
+          pip install pytest-reportlog pytest-cov parameterized datasets scipy einops
+          pip install "pytest>=7.2.0,<8.0.0" # see: https://github.com/huggingface/transformers/blob/ce4fff0be7f6464d713f7ac3e0bbaafbc6959ae5/setup.py#L148C6-L148C26
+
```
Comment (muellerzr) [inline .github/workflows/tests-pr.yml:41]: For sake of debugging, highly recommend adding a `pip freeze` step here

File: .github/workflows/tests-pr.yml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+      - uses: actions/checkout@v3
+      - name: Pip install and checkout
+        run: |
+          source activate peft
+          pip uninstall bitsandbytes
+          git remote add upstream https://github.com/TimDettmers/bitsandbytes.git
+          git fetch upstream pull/${{ github.event.inputs.pr_number }}/head:test-pr
+          git checkout test-pr
+          cmake -B . -DCOMPUTE_BACKEND=cuda -S . && cmake --build . && pip install -e . && pip freeze | grep bitsandbytes
+          pip install pytest-reportlog pytest-cov parameterized datasets scipy einops
```
Comment (muellerzr) [inline .github/workflows/tests-pr.yml]: Should these be added as test deps to the setup.py? So `pip install -e .[testing]` can be done in the prior line?

**Review by muellerzr — COMMENTED** [review]
Nice job! Very clean :) Just one or two nits

File: .github/workflows/tests-pr.yml (hunk 7 above)
Comment (younesbelkada) [reply .github/workflows/tests-pr.yml:41]: Makes sense ! done !

File: .github/workflows/tests-pr.yml (hunk 8 above)
Comment (younesbelkada) [reply .github/workflows/tests-pr.yml]: Makes sense, done !

**Review by muellerzr — APPROVED** [review]
Nice! I think you can use `gh workflow run` to actually test this before merging

**Comment by younesbelkada** [comment]
Thanks ! I think we need to enable the runners for this repo before doing so cc @glegendre01 🙏

File: .github/workflows/tests-pr.yml [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    inputs:
+      pr_number:
+        description: 'PR number'
+        required: true 
+        
+env:
+  # To be able to run tests on CUDA 12.2
+  NVIDIA_DISABLE_REQUIRE: "1"
+  SLACK_API_TOKEN: ${{ secrets.SLACK_API_TOKEN }}
+  OUTPUT_SLACK_CHANNEL_ID: "C071049S8EM"
```
Comment (younesbelkada) [inline .github/workflows/tests-pr.yml]: TODO: change to bnb colab

File: .github/workflows/tests-pr.yml [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+jobs:
+  run_tests_single_gpu:
+    strategy:
+      fail-fast: false
+    runs-on: [self-hosted, single-gpu, nvidia-gpu, t4, ci]
```
Comment (younesbelkada) [inline .github/workflows/tests-pr.yml:20]:
```suggestion
    runs-on: [self-hosted, single-gpu, nvidia-gpu, a10, ci]
```

**Pull request merged**